### PR TITLE
Fix dockerfile & add scripts

### PR DESCRIPTION
**Pull request by syinari0123**
## Environment
- Ubuntu 18.04
- GeForce GTX 1080
- Docker version 19.03.12, build 48a66213fe

## Issue about this pull request
Hi, I tried to run your DeepFactors on the docker environment.
However, on the docker environment created from your `Dockerfile.ubuntu`, I encountered an error when running demo file.
```sh
$ build/bin/df_demo \
>    --flagfile=data/flags/dataset_odom.flags \
>   --source_url=tum://~/datasets/tum_rgbd/rgbd_dataset_freiburg3_structure_notexture_near/

...

I0808 13:44:50.628271    11 deepfactors.cpp:183] Loading vocabulary from: data/voc/small_voc.yml.gz
I0808 13:44:50.667205    11 live_demo.cpp:129] Entering processing loop
I0808 13:44:50.667223    11 live_demo.cpp:140] Initializing system on the first frame
I0808 13:44:50.667253    50 live_demo.cpp:238] Visualization thread started
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X11 Error: BadMatch (invalid parameter attributes)
X11 Error: BadValue (integer parameter out of range for operation)
terminate called after throwing an instance of 'std::runtime_error'
  what():  Pangolin X11: Failed to create an OpenGL context
Aborted (core dumped)
```
In this pull request, I submit a modification of `Dockerfile.ubuntu`  to avoid this error!
In addition, I add scripts for building and running this docker file for convenience (`build_docker.sh` & `run_docker.sh`).

### 1. Dockerfile.ubuntu
- Related to https://github.com/jczarnowski/DeepFactors/issues/4, set `LD_LIBRARY_PATH` here. ([link](https://github.com/syinari0123/DeepFactors/blob/43116b4cbad3466da7472d48d955c74b9cb4e952/Dockerfile.ubuntu#L5))
```sh
# Add library path
ENV LD_LIBRARY_PATH ./thirdparty/install/lib:${LD_LIBRARY_PATH}
```
- To avoid X11 display error, I add this nvidia driver setting. ([link](https://github.com/syinari0123/DeepFactors/blob/43116b4cbad3466da7472d48d955c74b9cb4e952/Dockerfile.ubuntu#L7-9))
```
# Add nvidia driver settings
ENV NVIDIA_VISIBLE_DEVICES ${NVIDIA_VISIBLE_DEVICES:-all}
ENV NVIDIA_DRIVER_CAPABILITIES ${NVIDIA_DRIVER_CAPABILITIES:+$NVIDIA_DRIVER_CAPABILITIES,}graphics
```
- This code only supports, inference using pretrained weight. So I download the weight in this docker environment. ([link](https://github.com/syinari0123/DeepFactors/blob/43116b4cbad3466da7472d48d955c74b9cb4e952/Dockerfile.ubuntu#L33))
```sh
RUN set -x && bash ./scripts/download_network.bash
```

### 2. build_docker.sh
- Add a script to build a docker environment. ([link](https://github.com/syinari0123/DeepFactors/blob/43116b4cbad3466da7472d48d955c74b9cb4e952/scripts/docker/build_docker.sh#L1))
```sh
docker build -t deepfactors -f Dockerfile.ubuntu .
```
### 3. run_docker.sh
- Add a script to run docker image. ([link](https://github.com/syinari0123/DeepFactors/blob/43116b4cbad3466da7472d48d955c74b9cb4e952/scripts/docker/run_docker.sh#L1-9))
```sh
xhost +local:
sudo docker run -it \
    --rm \
    --runtime=nvidia \
    -v /home/$(id --user --name):/home/$(id --user --name) \
    -e DISPLAY=$DISPLAY \
    -v /tmp/.X11-unix/:/tmp/.X11-unix:ro \
    deepfactors
xhost -local:
```

## How to use these scripts to create a docker environment
In summary, you can create docker environment using these scripts as follows.
```sh
$ ./scipts/docker/build_docker.sh
$ ./scipts/docker/run_docker.sh
root@35bec0aa4aa2:/DeepFactors# build/bin/df_demo \
> --flagfile=data/flags/dataset_odom.flags \
> --source_url=tum://~/datasets/tum_rgbd/rgbd_dataset_freiburg3_structure_notexture_near/
 ```

### CODE REVIEW

**Comment by jczarnowski** [comment]
Hi, many thanks for the pull request!

I mostly use the Docker to test building on Ubuntu, and have only briefly played with running DeepFactors from inside the container. The contribution in this PR is definitely useful, as more people might want to use it only from Docker.

I have some comments regarding the changes:

1. It seems that running the `download_network.bash` script during docker build will cause the network to be re-downloaded every time someone makes changes to the code and wants to rebuild the image. It might be better to run the download script within `build_docker.sh` before building the image (weights will be copied with the code) and add a check to download the weights only if they dont exist already?

2. `run_docker.sh` 
**Home mount** -- choosing to mount the user home is a bit arbitrary as many people store datasets on external drives which are mounted under a different location. I am not sure how to fix this problem though. I can potentially see something like passing some parameter like <data_dir> to the `run_docker.sh` script which will mount that in docker.
**Running docker with sudo** -- technically it would be better to move the sudo call outside of the script, as people might not use sudo but rather rely on the user to be in the `docker` group. If you remove `sudo` from the script, the user can choose whether to execute the whole script with sudo or just as it is. This would cause problems with the home mount though as the reported username will be `root`.

Regardless of that, I think adding `run_docker.sh` even if the above is not fixed will be useful, just so people can use it as an example.

**Comment by syinari0123** [comment]
Thank you for your quick and appropriate comments!
I have made new modifications based on your suggestions.

1. `download_network.sh`: To avoid the weight's re-downloading in `build_docker.sh`, I decided to make a modification in  `download_network.sh`. Before downloading pretrained weights, this code checks if the weights have already been downloaded or not as follows. ([link](https://github.com/syinari0123/DeepFactors/blob/Fix-dockerfile-and-add-scripts/scripts/download_network.bash#L5))
```sh
#!/bin/bash
URL="https://www.dropbox.com/s/fj5dfq9hkqv2hm1/scannet256_32.tar.gz"
WEIGHT_PATH="data/nets"

if [[ ! -d ${WEIGHT_PATH} ]]; then
  # Download pretrained weights
  mkdir -p ${WEIGHT_PATH}
  wget -qO - ${URL} | tar xvz -C ${WEIGHT_PATH}
else
  echo "Pretrained weights have already exist in ${WEIGHT_PATH}!"
fi
```

2. `run_docker.sh`: 
    - First, I removed `sudo` when running the docker command as you pointed out. ([link](https://github.com/syinari0123/DeepFactors/blob/Fix-dockerfile-and-add-scripts/scripts/docker/run_docker.sh#L5)) 
    - And, to respond to your suggestion that it would not be a good idea that the mount directory is fixed as `usr/home`, I modified the code to show the mount directory (`DATA_PATH`) is changeable explicitly. Every user can change this to their environment. ([link](https://github.com/syinari0123/DeepFactors/blob/Fix-dockerfile-and-add-scripts/scripts/docker/run_docker.sh#L2))
```sh
# You can change this mount directory to your environment.
DATA_PATH="/home/$(id --user --name)"

xhost +local:
docker run -it \
    --rm \
    --runtime=nvidia \
    -v ${DATA_PATH}:${DATA_PATH} \
    -e DISPLAY=$DISPLAY \
    -v /tmp/.X11-unix/:/tmp/.X11-unix:ro \
    deepfactors
xhost -local:
```

**Comment by jczarnowski** [comment]
Amazing!
Thank you for taking the time to contribute back to the repo.

Merging

**Comment by syinari0123** [comment]
Thank you for the polite reviews to my sudden pull request. It was very helpful!

I'm looking forward to your next great work!
